core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -499,7 +500,10 @@ private[log] class Cleaner(val id: Int,
       val timestampType = firstMessageOffset.message.timestampType
       val messageWriter = new MessageWriter(math.min(math.max(MessageSet.messageSetSize(messages) / 2, 1024), 1 << 16))
       messageWriter.write(codec = compressionCodec, timestamp = magicAndTimestamp.timestamp, timestampType = timestampType, magicValue = messageFormatVersion) { outputStream =>
-        val output = new DataOutputStream(CompressionFactory(compressionCodec, outputStream))
+        val compressor = CompressionFactory(compressionCodec, outputStream)
+        if (compressionCodec == LZ4CompressionCodec && magicAndTimestamp.magic == Message.MagicValue_V0)
+          compressor.asInstanceOf[KafkaLZ4BlockOutputStream].useBrokenHC = true // Maintain compatibility with 0.8 / 0.9 clients
```

Comment:
I disagree -- you're inviting other compression codecs to couple themselves with message format, and that is not the intent of this change.

Nonetheless, I'll change it since ultimately I am not the one that has to maintain this codebase :)